apps/src/feedback.js

```diff
@@ -967,52 +793,14 @@ FeedbackUtils.prototype.getTestResults = function(levelComplete, requiredBlocks,
         this.studioApp_.TestResults.ALL_PASS :
         this.studioApp_.TestResults.TOO_FEW_BLOCKS_FAIL;
   }
-  if (shouldCheckForEmptyBlocks) {
-    var emptyBlockFailure = this.checkForEmptyContainerBlockFailure_();
-    if (emptyBlockFailure !== TestResults.ALL_PASS) {
-      return emptyBlockFailure;
-    }
-  }
-  if (!options.allowTopBlocks && this.hasExtraTopBlocks()) {
-    return TestResults.EXTRA_TOP_BLOCKS_FAIL;
-  }
-  if (Blockly.useContractEditor || Blockly.useModalFunctionEditor) {
-    if (this.hasUnusedParam_()) {
-      return TestResults.UNUSED_PARAM;
-    }
-    if (this.hasUnusedFunction_()) {
-      return TestResults.UNUSED_FUNCTION;
-    }
-    if (this.hasParamInputUnattached_()) {
-      return TestResults.PARAM_INPUT_UNATTACHED;
-    }
-    if (this.hasIncompleteBlockInFunction_()) {
-      return TestResults.INCOMPLETE_BLOCK_IN_FUNCTION;
-    }
-  }
-  if (this.hasQuestionMarksInNumberField_()) {
-    return TestResults.QUESTION_MARKS_IN_NUMBER_FIELD;
-  }
-  if (!this.hasAllRequiredBlocks_(requiredBlocks)) {
-    return levelComplete ?
-        TestResults.MISSING_BLOCK_FINISHED :
-        TestResults.MISSING_BLOCK_UNFINISHED;
-  }
-  var numEnabledBlocks = this.getNumCountableBlocks();
-  if (!levelComplete) {
-    if (this.studioApp_.IDEAL_BLOCK_NUM &&
-        this.studioApp_.IDEAL_BLOCK_NUM !== Infinity &&
-        numEnabledBlocks < this.studioApp_.IDEAL_BLOCK_NUM) {
-      return TestResults.TOO_FEW_BLOCKS_FAIL;
-    }
-    return TestResults.LEVEL_INCOMPLETE_FAIL;
-  }
-  if (this.studioApp_.IDEAL_BLOCK_NUM &&
-      numEnabledBlocks > this.studioApp_.IDEAL_BLOCK_NUM) {
-    return TestResults.TOO_MANY_BLOCKS_FAIL;
-  } else {
-    return TestResults.ALL_PASS;
-  }
+
+  // If we get this far, we assume that Blockly is being used.
+  var blockStaticAnalyzer = new BlockStaticAnalyzer(Blockly);
+  blockStaticAnalyzer.setShouldCheckForEmptyBlocks(shouldCheckForEmptyBlocks);
+  blockStaticAnalyzer.setAllowExtraTopBlocks(options.allowTopBlocks);
+  blockStaticAnalyzer.setIdealBlockCount(this.studioApp_.IDEAL_BLOCK_NUM);
+  blockStaticAnalyzer.setRequiredBlocks(requiredBlocks);
+  return blockStaticAnalyzer.runStaticAnalysis(levelComplete);
```

Comment:
As I think about this more, I'm not fully convinced this should be a class. It seems like the pattern is create an object with some options, call a single method on the object, and throw the object away. Is there an argument for doing it this way instead of having BlockStaticAnalyzer have a set of static methods, each of which takes a Blockly and a set of options.